simulation.py

```diff
@@ -36,6 +34,64 @@ def update_electric_field(self):
 		self.E_vector = electric_field_function(
 			self._source_position, self._source_velocity, self._source_acceleration, self._position)
 
+class SimulationEngine(object):
+	def __init__(self, charge_motion, direction_unit_vectors):
+		self._charge_motion = charge_motion
+		#directions that field samples move in
+		self._direction_unit_vectors = direction_unit_vectors
+
+		self._current_time = 0.0
+		self._field_samples = []
+
+		self._update_charge_kinematics()
+
+	def _update_charge_kinematics(self):
+		charge_kinematics = self._charge_motion.get_state(self._current_time)
+
+		self._charge_position = np.array(charge_kinematics.position)
+		self._charge_velocity = np.array(charge_kinematics.velocity)
+		self._charge_acceleration = np.array(charge_kinematics.acceleration)
+
+	def _emit_new_field_samples(self):
+		for unit_vec in self._direction_unit_vectors:
+			self._field_samples.append(ElectricFieldSample(unit_vec, self._charge_position, 
+				self._charge_velocity, self._charge_acceleration))
+
+	def advance_simulation(self, dt):
+		#emit new samples from the old charge position, we must do this before updating the position
+		self._emit_new_field_samples()
+
+		self._current_time += dt
+		self._update_charge_kinematics()
+
+		late_index = None
+		curr_index = 0
+		
+		for field_sample in self._field_samples:
+			field_sample.advance(dt)
+
+			# we need to delete the first few samples as they are too old, they are ordered chronologically
+			# by creation time. These old are so far you can't see them so don't process them to save
+			# CPU and RAM
+			if field_sample._time > MAX_DISTANCE / SPEED_OF_LIGHT:
+				late_index = curr_index
+			curr_index += 1
+
+		if late_index is not None:
+			#TODO: self._field_samples should be a deque and we can pop
+			del self._field_samples[:late_index]
+
+	def get_next_frames(self, dt, frame_count):
```

Comment:
Could consider using `@property` rather than getter?

Reply:
I did this for the other get methods, but not this one because it will be more complicated (it's not yet implemented, will plan to implement this method in another PR. It will basically call advance_simulation(dt) a multiple number of times (frame_count number of times) and return a large list of (time, field_samples_list). The idea is to give the renderer a large list of future frames to render.